scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/license/LicenseSchedulingPolicy.java

```diff
@@ -0,0 +1,142 @@
+ * as published by the Free Software Foundation: version 3 of
+ * the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.scheduler.policy.license;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Properties;
+
+import org.apache.log4j.Logger;
+import org.ow2.proactive.scheduler.common.task.TaskStatus;
+import org.ow2.proactive.scheduler.descriptor.EligibleTaskDescriptor;
+import org.ow2.proactive.scheduler.descriptor.EligibleTaskDescriptorImpl;
+import org.ow2.proactive.scheduler.policy.ExtendedSchedulerPolicy;
+import org.ow2.proactive.utils.NodeSet;
+
+
+/**
+ * 
+ * This Policy is designed to manage software licenses.
+ * When a task contains the generic information REQUIRE_LICENSES,
+ * this policy will check if there is an available license per
+ * software to return true
+ *
+ */
+public class LicenseSchedulingPolicy extends ExtendedSchedulerPolicy {
+
+    private static final Logger logger = Logger.getLogger(LicenseSchedulingPolicy.class);
+
+    public static final String REQUIRED_LICENSES = "REQUIRED_LICENSES";
+
+    // A map of fixed size lists. A map entry per software license and a list per map entry to handle tasks using this software license.
+    private Map<String, FixedSizeArrayList> eligibleTasksDescriptorsLicenses = null;
+
+    @Override
+    public boolean isTaskExecutable(NodeSet selectedNodes, EligibleTaskDescriptor task) {
+
+        logger.debug("Selected Nodes: " + selectedNodes);
+
+        logger.debug("Analysing task: " + ((EligibleTaskDescriptorImpl) task).getInternal().getName());
+
+        // Retrieve required licenses names from the task generic informations
+        final String requiredLicenses = ((EligibleTaskDescriptorImpl) task).getInternal()
+                                                                           .getRuntimeGenericInformation()
+                                                                           .get(REQUIRED_LICENSES);
+
+        // If it requires software licenses
+        if (requiredLicenses != null) {
+
+            // Initialize eligibleTasksDescriptorsLicenses
+            if (eligibleTasksDescriptorsLicenses == null)
+                eligibleTasksDescriptorsLicenses = new HashMap<String, FixedSizeArrayList>();
+
+            // Retrieve the maximum licenses numbers per software from the configuration file
+            final Properties properties = LicenseConfiguration.getConfiguration().getProperties();
```

Comment:
configuration should not be read for each task. If you have 10000 tasks to evaluate, it would be a huge overhead.

Configuration should be done only once. We can accept for now that the scheduler must be restarted if the configuration changes.